server/questions.js

```diff
@@ -544,4 +544,62 @@ exports.voteReply = function (replyId, vote, userId, callback) {
             return callback('Could not submit the vote on the reply', null);
         }
     });
+}
+
+exports.isUserLocked = function(userId, question, callback){
+    var lastSubmissionTime;
+    var currentDate = common.getDateObject();
+
+    // check if user already got the question correct
+    for (var obj = 0; obj < question.correctAttempts.length; obj ++){
+        if(question.correctAttempts[obj]['userId'] == userId){
+            return callback(false,null);
+        }
+    }
+    // find the lastSubmissionTime for this user
+    for (var obj = 0; obj < question.userSubmissionTime.length; obj++){
+        if(question.userSubmissionTime[obj]['userId'] === userId){
+            lastSubmissionTime = question.userSubmissionTime[obj]['submissionTime'];
+        }
+    }
+
+    if(typeof lastSubmissionTime !== 'undefined'){
+        var diff = Math.abs(currentDate - lastSubmissionTime);
+        if (diff < common.waiting_time){
+            return callback(true,common.getTime(common.waiting_time - diff));
+        }
+    }
+    return callback(false,null);
+}
+
+/*Updates the Users Submission time for a Question*/
+exports.updateUserSubmissionTime = function(userId, question, callback){
+    var query = {_id:question._id};
+    var update = {};
+    var currentDate = common.getDateObject();
+    var userInList = false;
+    // find the user in the submission list
+    for (var obj = 0; obj < question.userSubmissionTime.length; obj++){
+        if(question.userSubmissionTime[obj]['userId'] === userId){
+            userInList = true;
```

Comment:
If we find the user early on in the list, we are still looping through till the end, this is inefficient. Please take a look at the JavaScript function `find()`, could be helpful to find the first item that satisfies this criteria.

Reply:
can't use find() since it needs to find the exact object which is not possible. added break to not loop over.

Reply:
Sure, I think it's possible, I would recommend you take another look into it but that's up to you